src/main/java/com/gregtechceu/gtceu/api/gui/widget/TankWidget.java

```diff
@@ -567,23 +566,31 @@ private int tryClickContainer(boolean isShiftKeyDown) {
         if (handler == null) return -1;
         int maxAttempts = isShiftKeyDown ? currentStack.getCount() : 1;
         FluidStack initialFluid = fluidTank.getFluidInTank(tank);
+        FluidStack initialFluidCopy = initialFluid.copy();
         if (allowClickFilled && initialFluid.getAmount() > 0) {
             boolean performedFill = false;
+            ItemStack filledResult = ItemStack.EMPTY;
             for (int i = 0; i < maxAttempts; i++) {
                 FluidActionResult result = FluidUtil.tryFillContainer(currentStack, fluidTank, Integer.MAX_VALUE, null,
                         false);
                 if (!result.isSuccess()) break;
                 ItemStack remainingStack = FluidUtil
                         .tryFillContainer(currentStack, fluidTank, Integer.MAX_VALUE, null, true).getResult();
-                currentStack.shrink(1);
                 performedFill = true;
-                if (!remainingStack.isEmpty() && !player.addItem(remainingStack)) {
-                    Block.popResource(player.level(), player.getOnPos(), remainingStack);
-                    break;
+
+                currentStack.shrink(1);
+
+                if (filledResult.isEmpty())
+                    filledResult = remainingStack.copy();
+                else if (ItemStack.isSameItemSameTags(filledResult, remainingStack))
+                    filledResult.grow(1);
+                else {
+                    player.getInventory().placeItemBackInInventory(filledResult);
+                    filledResult = remainingStack.copy();
```

Comment:
Last branch of if/else chain has brackets, so all branches should have brackets for clarity